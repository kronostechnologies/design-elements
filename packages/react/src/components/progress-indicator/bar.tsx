import { ReactText, VoidFunctionComponent } from 'react';
import styled from 'styled-components';

interface BarProps {
    className?: string;
    color: string;
    percent: number;
    endLabel: ReactText;
}

const Container = styled.div`
    p {
        color: ${({ theme }) => theme.component['progress-indicator-label-text-color']};
        letter-spacing: 0.02875rem;
        line-height: 1.5rem;
        margin: 0;
        text-align: right;
    }
`;

const Progress = styled.div`
    background-color: ${({ theme }) => theme.component['progress-indicator-empty-track-color']};
    border-radius: var(--border-radius);
    height: 0.5rem;
    margin-bottom: var(--spacing-half);
    width: 100%;
`;

const StyledBar = styled.div<Omit<BarProps, 'endLabel'>>`
    background: ${({ color }) => color};
    border-radius: var(--border-radius);
    height: 0.5rem;
    width: ${({ percent }) => Math.min(Math.max(percent, 0), 100)}%;
`;

export const Bar: VoidFunctionComponent<BarProps> = ({
    className, color, percent, endLabel,
}) => (
    <Container className={className}>
        <Progress>
            <StyledBar color={color} percent={percent} />
        </Progress>
        <p>{endLabel}</p>
    </Container>
);
