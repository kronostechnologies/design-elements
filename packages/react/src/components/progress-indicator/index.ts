export { ProgressIndicator } from './progress-indicator';
export type { ProgressIndicatorProps } from './progress-indicator';
