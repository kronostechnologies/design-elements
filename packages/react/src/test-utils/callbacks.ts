export function doNothing(): void {
    // do nothing
}
