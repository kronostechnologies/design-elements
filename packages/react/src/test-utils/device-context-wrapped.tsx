import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
