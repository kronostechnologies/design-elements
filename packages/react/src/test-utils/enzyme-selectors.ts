import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
