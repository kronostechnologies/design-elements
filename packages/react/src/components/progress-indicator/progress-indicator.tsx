import { ReactText, VoidFunctionComponent } from 'react';
import styled from 'styled-components';
import { Theme } from '../../themes';
import { Bar } from '../bar/bar';

const Label = styled.label`
    color: ${(props: { theme: Theme }) => props.theme.ref['color-black']};
    display: block;
    font-size: 0.875rem;
    letter-spacing: 0.02875rem;
    line-height: 1.5rem;
    margin-bottom: var(--spacing-half);
`;

const Container = styled.div`
    & + & {
        margin-top: var(--spacing-half);
    }
`;

interface Props {
    className?: string;
    /** [0 - 100] */
    percent: number;
    color: string;
    /** End label */
    resultLabel: ReactText;
    /** Top label */
    descriptionLabel?: string;
}

export const ProgressIndicator: VoidFunctionComponent<Props> = ({
    className, color, descriptionLabel, resultLabel, percent,
}) => (
    <Container className={className}>
        <Label>{descriptionLabel}</Label>
        <Bar
            color={color}
            endLabel={resultLabel}
            percent={percent}
        />
    </Container>
);
