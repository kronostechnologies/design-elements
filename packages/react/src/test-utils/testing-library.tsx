import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { DesignSystem, DesignSystemProps } from '../components/design-system';
import { equisoftTheme } from '../theme';

export function renderWithProviders(
    ui: Parameters<typeof render>[0],
    renderOptions: Omit<Parameters<typeof render>[1], 'wrapper'> | undefined = undefined,
    wrapperProps: DesignSystemProps = {},
): ReturnType<typeof render> {
    const {
        isolateStyles,
        language,
        staticDevice,
        theme,
    } = {
        isolateStyles: false,
        language: 'fr',
        theme: equisoftTheme,
        ...wrapperProps,
    };
    return render(
        ui,
        {
            ...renderOptions,
            wrapper: ({ children }) => (
                <MemoryRouter future={{ v7_startTransition: false, v7_relativeSplatPath: false }}>
                    <DesignSystem
                        isolateStyles={isolateStyles}
                        language={language}
                        staticDevice={staticDevice}
                        theme={theme}
                    >
                        {children}
                    </DesignSystem>
                </MemoryRouter>
            ),
        },
    );
}
