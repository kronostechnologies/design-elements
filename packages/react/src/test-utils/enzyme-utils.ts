import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

export async function waitForComponentToPaint<P = {}>(
    wrapper: ReactWrapper<P>,
    amount = 0,
): Promise<void> {
    await act(async () => {
        await new Promise((resolve) => setTimeout(resolve, amount));
        wrapper.update();
    });
}

export function expectFocusToBeOn(element: ReactWrapper): void {
    expect(document.activeElement).toBe(element.getDOMNode());
}
