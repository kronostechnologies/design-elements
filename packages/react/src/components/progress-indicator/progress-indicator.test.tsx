import { renderWithProviders } from '../../test-utils/testing-library';
import { ProgressIndicator } from './progress-indicator';

describe('ProgressIndicator', () => {
    it('matches snapshot', () => {
        const { asFragment } = renderWithProviders(<ProgressIndicator />);

        expect(asFragment()).toMatchSnapshot();
    });
});
