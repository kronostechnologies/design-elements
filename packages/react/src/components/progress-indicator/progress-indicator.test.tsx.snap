// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProgressIndicator matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  color: #006296;
  height: 64px;
  width: 64px;
}

<svg
    class="c0"
  />
</DocumentFragment>
`;
